src/viem/saga.ts

```diff
@@ -1,35 +1,238 @@
+    const simulateContractMethod = yield* call(getSimulateContractMethod, {
+      wallet,
+      tokenAddress,
+      amount,
+      recipientAddress,
+      comment,
+      feeInfo,
+    })
+
+    // Explicit type is specified here. The actual return type of this function is
+    // something like
+    // SimulateContractReturnType<typeof stableToken.abi, 'transferWithComment'>
+    // | SimulateContractReturnType<typeof erc20.abi, 'transfer'>
+    // but TSC doesn't like it when passed to writeContract
+    const { request }: SimulateContractReturnType = yield* call(simulateContractMethod)
+
+    const hash = yield* call([wallet, 'writeContract'], request)
+    Logger.debug(TAG, 'Transaction successfully submitted. Hash:', hash)
+
+    // TODO: wait for receipt with a timeout, confirm tx and return
+
+    return hash
+  } catch (err) {
+    Logger.warn(TAG, 'Transaction failed', err)
+    throw err
+  }
+}
+
+/**
+ * Gets a function that invokes simulateContract for the appropriate contract
+ * method based on the token. If the token is a stable token, it uses the
+ * `transferWithComment` on the stable token contract, otherwise the `transfer`
+ * method on the ERC20 contract
+ *
+ * @param options an object containing the arguments
+ * @returns a function that invokes the simulateContract method
+ */
+function* getSimulateContractMethod({
+  wallet,
+  tokenAddress,
+  amount,
+  recipientAddress,
+  comment,
+  feeInfo,
+}: {
+  wallet: ViemWallet
+  recipientAddress: string
+  amount: BigNumber
+  tokenAddress: string
+  comment: string
+  feeInfo: FeeInfo
+}) {
+  if (!wallet.account) {
+    // this should never happen
+    throw new Error('no account found in the wallet')
+  }
+
+  const tokenInfo = yield* call(getTokenInfo, tokenAddress)
+
+  const convertedAmount = BigInt(yield* call(tokenAmountInSmallestUnit, amount, tokenAddress))
+
+  if (isStablecoin(tokenInfo)) {
+    const userAddress = wallet.account.address
+    const encryptedComment = yield* call(
+      encryptComment,
+      comment,
+      recipientAddress,
+      userAddress,
+      true
+    )
+
+    const { feeCurrency, gas, maxFeePerGas } = yield* call(getSendTxFeeDetails, {
+      recipientAddress,
+      amount,
+      tokenAddress,
+      feeInfo,
+      encryptedComment: encryptedComment || '',
+    })
+
+    Logger.debug(TAG, 'Calling simulate contract for transferWithComment', {
+      recipientAddress,
+      convertedAmount,
+      feeCurrency,
+      gas,
+      maxFeePerGas,
+    })
+
+    return () =>
+      publicClient.celo.simulateContract({
+        address: getAddress(tokenAddress),
+        abi: stableToken.abi,
+        functionName: 'transferWithComment',
+        account: wallet.account,
+        args: [getAddress(recipientAddress), convertedAmount, encryptedComment || ''],
+        feeCurrency,
+        gas,
+        maxFeePerGas,
+      })
+  }
+
+  const { feeCurrency, gas, maxFeePerGas } = yield* call(getSendTxFeeDetails, {
```

Comment:
Does passing `''` for `encryptedComment` here fundamentally change the results of this call? In other words, rather than having two calls to `getSendTxFeeDetails`, can we pull this out to a single call above the `if (isStablecoin(tokenInfo))` block above? I guess this complicates things further though, since we need to encrypt the comment... could we do something like the following above the `if` statement?

```ts
const encryptedComment = comment ? yield* call(
      encryptComment,
      comment,
      recipientAddress,
      userAddress: wallet.account.address,
      true
) : ''

const { feeCurrency, gas, maxFeePerGas } = yield* call(getSendTxFeeDetails, {
      recipientAddress,
      amount,
      tokenAddress,
      feeInfo,
      encryptedComment
})
```

Reply:
I've simplified this. Calling fee details with an empty comment should be fine for non stables. It will be ignored when a send tx is built as it conditionally uses the stable token or erc20 contract.